### PR TITLE
Flatten expressions commutatively

### LINKED ISSUE #92: a + 1 - 1 doesn't simplify
Should go to a, goes to a + 1 - 1.

### PR DESCRIPTION
**Pull request by ayazhafiz**
This commit adds an option to flatten/fold expressions a little bit smarter
than we currently do by flattening and folding an expression
commutatively.

Not sure if there should be a better name, like `flat_fold` or similar?
Or we can rename the current usages of `flatten` and `unflatten`, which
don't do any expression folding, to `rollup` and `unroll` or similar.
Feedback wanted on this.

Closes #92

### CODE REVIEW

**Comment by ayazhafiz** [comment]
I'll add documentation in a subsequent pass, first want to get feedback if this implementation seems reasonable.

File: libslide/src/grammar.rs [hunk 1]
```diff
@@ -365,6 +382,21 @@ pub struct UnaryExpr<E: Expression> {
...
+where
+    E: Expression,
+{
+    pub fn negate<T>(expr: T) -> Self
+    where
+        T: Into<Rc<E>>,
+    {
+        Self {
+            op: UnaryOperator::SignNegative,
+            rhs: expr.into(),
```
Comment (luke-bhan) [inline libslide/src/grammar.rs:395]: shouldn't negate not always be sign negative. if you negating a negative it should flip

File: libslide/src/partial_evaluator.rs [hunk 2]
```diff
@@ -134,6 +138,8 @@ mod tests {
         unwrap_brackets_nested:           "[a] + [1]" => "a + 1"
 
         flattened_addition:             "1 + 2 - b + 3 - b" => "6 - b - b"
+
+        issue_92: "a + 1 - 1" => "a"
     }
```
Comment (luke-bhan) [inline libslide/src/partial_evaluator.rs:143]: No. We will not be naming these things by issue number- especially rules. Call it literal_cancellation or something like that.

File: libslide/src/partial_evaluator/flatten.rs [hunk 3]
```diff
@@ -0,0 +1,139 @@
+use crate::grammar::*;
+use crate::utils::{unflatten_binary_expr, UnflattenStrategy};
+
+use std::collections::{HashMap, VecDeque};
+use std::rc::Rc;
+
+pub fn flatten_expr(expr: &Rc<Expr>) -> Rc<Expr> {
+    match expr.as_ref() {
```
Comment (luke-bhan) [inline libslide/src/partial_evaluator/flatten.rs:8]: why is expr getting match as_ref. shouldn't it already be ref due to the fact you're borrowing it.

File: libslide/src/partial_evaluator/flatten.rs [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        // TODO: handle everything else
+        _ => Rc::clone(expr),
+    }
+}
+
+pub fn flatten_add_or_sub(o_lhs: &Rc<Expr>, o_rhs: &Rc<Expr>, is_subtract: bool) -> Rc<Expr> {
+    let lhs = flatten_expr(o_lhs);
+    let rhs = flatten_expr(o_rhs);
+
+    let mut coeff = 0f64;
```
Comment (luke-bhan) [inline libslide/src/partial_evaluator/flatten.rs:25]: this is a nitpick but we should stay consistnent.. let mut coeff: f64 = 0.0;

File: libslide/src/partial_evaluator/flatten.rs [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    let mut terms = HashMap::<&Rc<Expr>, f64>::new();
+    let mut args = VecDeque::with_capacity(2);
+    let base_args = [lhs, rhs];
+    args.extend(base_args.iter());
+    let mut args_before_neg = 1;
```
Comment (luke-bhan) [inline libslide/src/partial_evaluator/flatten.rs:30]: this deque could probably be created in one line;
I think rc should be an empty pointer, but Idk if default is implemented off my head
let mut args: VecDeque = vec![Rc<Expr>::default(), Rc<Expr>::default(), lhs, rhs]

File: libslide/src/partial_evaluator/flatten.rs [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+            }
+        }
+    }
+
+    let mut new_args: Vec<Rc<Expr>> = Vec::with_capacity(1 + terms.len());
+    if coeff != 0. {
+        new_args.push(Rc::from(Expr::Const(coeff)));
+    }
+    for (term, coeff) in terms.into_iter() {
+        if coeff == 0. {
```
Comment (luke-bhan) [inline libslide/src/partial_evaluator/flatten.rs:74]: you don't need into_iter() here.

File: libslide/src/partial_evaluator/types.rs [hunk 7]
```diff
@@ -2,12 +2,14 @@ use crate::evaluator_rules::RuleName;
...
 }
 
 impl Default for EvaluatorContext {
     fn default() -> Self {
         Self {
             rule_blacklist: vec![],
+            always_flatten: true,
         }
     }
 }
```
Comment (luke-bhan) [inline libslide/src/partial_evaluator/types.rs:15]: I think this default constructor should take a parameter to know to flatten rather then set to true but this can be down in another PR.

File: libslide/src/partial_evaluator/flatten.rs [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+        "1 - 1 -> 0",
+        "1 + 2 - 3 -> 0",
+        "1 - 2 + 3 -> 2",
+        "a - a + 1 -> 1",
+        "a + 1 - 1 -> a",
+    ];
+
+    #[test]
+    fn flatten_cases() {
+        for case in CASES.into_iter() {
```
Comment (luke-bhan) [inline libslide/src/partial_evaluator/flatten.rs:129]: Again, do you need the into_iter() here. I don't think so :)

File: libslide/src/partial_evaluator/flatten.rs [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+            let mult = BinaryExpr::mult(Expr::Const(coeff), Rc::clone(term));
+            let expr: Expr = mult.into();
+            new_args.push(Rc::from(expr));
+        }
+    }
+
+    match new_args.len() {
+        0 => Rc::from(Expr::Const(0.)),
+        1 => new_args.into_iter().next().unwrap(),
+        _ => unflatten_binary_expr(&new_args, BinaryOperator::Plus, UnflattenStrategy::Left),
```
Comment (luke-bhan) [inline libslide/src/partial_evaluator/flatten.rs:91]: why don't you just use get instead of casting to an iter and unwrapping in new_args

File: libslide/src/grammar.rs (hunk 1 above)
Comment (ayazhafiz) [reply libslide/src/grammar.rs:395]: the contructor is just to negate the given expression. I don't think this method should be responsible for unfolding the underlying expression to figure out if it needs to be normalized.

File: libslide/src/partial_evaluator.rs (hunk 2 above)
Comment (ayazhafiz) [reply libslide/src/partial_evaluator.rs:143]: Why not? These tests are mostly for regressions anyway. If we have a similar issue for something like `a + b + 9 + c -e -9` where 9 - 9 doesn't go to zero, what do we name this? `literal_cancellation_2` is fine but it leaves less context about when and where the bug appeared.

Related to this, I think we should move all tests into some kind of corpus (it can even just be a giant text document), similar to how rust/rustfmt do it. What do you think?

File: libslide/src/partial_evaluator/flatten.rs (hunk 3 above)
Comment (ayazhafiz) [reply libslide/src/partial_evaluator/flatten.rs:8]: This gets a ref to the underlying `Expr`

File: libslide/src/partial_evaluator/flatten.rs (hunk 4 above)
Comment (ayazhafiz) [reply libslide/src/partial_evaluator/flatten.rs:25]: good call, thanks

File: libslide/src/partial_evaluator/flatten.rs (hunk 5 above)
Comment (ayazhafiz) [reply libslide/src/partial_evaluator/flatten.rs:30]: `with_capacity` reserves memory but doesn't populate anything. There is no macro for creating vecdeques in the standard library, but if you want you can roll your own: https://github.com/bluss/maplit/pull/31/files. That can be in a subsequent PR though.

File: libslide/src/partial_evaluator/types.rs (hunk 7 above)
Comment (ayazhafiz) [reply libslide/src/partial_evaluator/types.rs:15]: The default ctor takes no parameters because it is the "default". There is a builder method to set this flag in the EvaluatorContext impl.

File: libslide/src/partial_evaluator/flatten.rs (hunk 5 above)
Comment (luke-bhan) [reply libslide/src/partial_evaluator/flatten.rs:30]: No, but you can just make a vector that easily casts into a deque. That’s how they do it on Rust docs

File: libslide/src/partial_evaluator/flatten.rs (hunk 9 above)
Comment (ayazhafiz) [reply libslide/src/partial_evaluator/flatten.rs:91]: `get` returns a reference, but we don't want that, we want to return the value, so it's find to destroy the vec

Comment (luke-bhan) [reply libslide/src/partial_evaluator/flatten.rs:91]: Getting a reference and casting it is faster then casting iter element and then getting a next and then unwrapping though

Comment (luke-bhan) [reply libslide/src/partial_evaluator/flatten.rs:91]: Getting a reference and casting it is faster then casting iter element and then getting a next and then unwrapping though

File: libslide/src/partial_evaluator/flatten.rs (hunk 5 above)
Comment (ayazhafiz) [reply libslide/src/partial_evaluator/flatten.rs:30]: Can you point me to documentation that does this? I'm not aware of anyway to do this

File: libslide/src/partial_evaluator/flatten.rs (hunk 9 above)
Comment (ayazhafiz) [reply libslide/src/partial_evaluator/flatten.rs:91]: wait what? casting how? We need to go from a reference to a raw value

File: libslide/src/partial_evaluator/flatten.rs (hunk 5 above)
Comment (luke-bhan) [reply libslide/src/partial_evaluator/flatten.rs:30]: https://doc.rust-lang.org/std/collections/struct.VecDeque.html
Reserve function shows them making one from a vec

File: libslide/src/partial_evaluator/flatten.rs (hunk 9 above)
Comment (luke-bhan) [reply libslide/src/partial_evaluator/flatten.rs:91]: I thought you can just dereference

File: libslide/src/partial_evaluator/flatten.rs (hunk 5 above)
Comment (ayazhafiz) [reply libslide/src/partial_evaluator/flatten.rs:30]: hmm..  i personally would rather keep it as is because I don't see a need to create and collect an iterator again. What do you think?

File: libslide/src/partial_evaluator/flatten.rs (hunk 9 above)
Comment (ayazhafiz) [reply libslide/src/partial_evaluator/flatten.rs:91]: No cuz that would move the value out of the vector which is unsound.

File: libslide/src/partial_evaluator/flatten.rs (hunk 5 above)
Comment (luke-bhan) [reply libslide/src/partial_evaluator/flatten.rs:30]: That’s fine

File: libslide/src/partial_evaluator/flatten.rs (hunk 9 above)
Comment (luke-bhan) [reply libslide/src/partial_evaluator/flatten.rs:91]: Why is it unsound if your destroying the deque anyway

Comment (ayazhafiz) [reply libslide/src/partial_evaluator/flatten.rs:91]: yeah but the compiler isn't smart enough to know that: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=47cedd7a1acf53f5bff12a0461a41f74

Comment (ayazhafiz) [reply libslide/src/partial_evaluator/flatten.rs:91]: There is a better way to do this though, you're right: remove the value from the vector with `swap_remove`

**Comment by ayazhafiz** [comment]
shoot I forgot to add docs!
